Multi-dimentional output strategy for DeepONet in PyTorch

kuangdai opened the pull request:
# How this was done

I made it in two steps:
1) moving the strategy classes from `tensorflow`'s `deeponet.py` to a separate file `deeponet_strategy.py` so that all backends can share it
2) adding strategeis to `pytorch` and `paddle` following the same way (with few synatx adaptions). Changes for `paddle` comes only after this PR is merged.

# Remaining Issue
I noted that in `call(self, x_func, x_loc, training=False)` of `DeepONetStrategy`, the argument `training` is completely unused. It is kept in this PR because I don't know if it serves some other purpose. I suggest removal of it.

# Tests
A file `test_strat.py` (renamed to `test_start.txt` for uploading) is attached in this description. It tests all the model classes (`DeepONet`, `DeepONetCartesianProd`, `PODDeepONet`) and strategy settings (single, independent, split both, split branch, split trunk), making sure that the forward pass yields the correct output shape. To reproduce the tests, put this file under the top directory and run

```bash
DDE_BACKEND=tensorflow python test_strat.py 
DDE_BACKEND=pytorch python test_strat.py 
DDE_BACKEND=paddle python test_strat.py
```

And you should see no errors.

[test_strat.txt](https://github.com/lululxvi/deepxde/files/13999388/test_strat.txt)

lululxvi commented on deepxde/nn/tensorflow/deeponet.py:
```diff
@@ -1,184 +1,19 @@
+    SplitBothStrategy,
+    SplitBranchStrategy,
+    SplitTrunkStrategy,
+)
```
```python
from .. import activations
from ..deeponet_strategy import (
    SingleOutputStrategy,
    IndependentStrategy,
    SplitBothStrategy,
    SplitBranchStrategy,
    SplitTrunkStrategy,
)
from ... import config
from ...backend import tf
```

lululxvi replied:
@kuangdai Reorder the import.

lululxvi commented:
If you make pytorch code as a separate PR, then I can merge this Tensorflow one right away.

kuangdai commented:
> If you make pytorch code as a separate PR, then I can merge this Tensorflow one right away.

I would suggest not rewinding the commits, as it may cause other issues. The change on tensorflow is simply separating the code into two, which does not serve a standalone purpose for a PR. Please take your time to check the pytorch implementation. No rush.

lululxvi commented on deepxde/nn/pytorch/deeponet.py:
```diff
@@ -1,24 +1,65 @@
+    SplitTrunkStrategy,
+)
+from ... import config
+from ...backend import torch
```
```python
from itertools import cycle

import torch

from .fnn import FNN
...
```

No need to import torch from backend. The reason of importing tf from backend is that tf has v1 and v2, which are managed by backend.

lululxvi commented on deepxde/nn/pytorch/deeponet.py:
```diff
@@ -27,40 +68,76 @@ def __init__(
+        return torch.concat(ys, dim=1)
 
-    def forward(self, inputs):
+    def forward(self, inputs, training=False):
```
pytorch doesn't use `training=False`, which is only needed in tf.

kuangdai replied:
Good spot. As I wrote in PR description, `training` is unused in `DeepONetStrategy.call()`. Shall we remove it from strategies?

lululxvi replied:
Yes, we can remove.

lululxvi commented on deepxde/nn/pytorch/deeponet.py:
```diff
@@ -1,24 +1,65 @@
-import torch
+__all__ = ["DeepONet", "DeepONetCartesianProd", "PODDeepONet"]
+
+from itertools import cycle
```
```
from itertools import cycle

import torch

from .fnn import FNN
```

kuangdai commented:
I removed argument `training`. 

Also, I realised that using `cycle` to iterate the biases inside the class is thread-unsafe, that is, if the external loop over `num_outputs` from strategies is multi-threaded, the cycle won't keep synchronised with `i` because multi-threading is unordered. Although no one will do this, it is better to have right logic. I changed to explicitly passing the index. This also makes the code more readable.

lululxvi commented:
> Also, I realised that using `cycle` to iterate the biases inside the class is thread-unsafe, that is, if the external loop over `num_outputs` from strategies is multi-threaded, the cycle won't keep synchronised with `i` because multi-threading is unordered. Although no one will do this, it is better to have right logic. I changed to explicitly passing the index. This also makes the code more readable.

Good point.

lululxvi commented on deepxde/nn/pytorch/deeponet.py:
```diff
@@ -160,42 +292,46 @@ def __init__(
         activation,
         kernel_initializer,
         layer_sizes_trunk=None,
-        regularization=None,
```
Keep this `regularization`, which may be needed in some cases.

lululxvi replied:
I was not clear. Only keep the `regularization` in POD-DeepONet.
